File: Google/146 LRU Cache.js
class LRUCache {
    constructor(capacity) {
        this.cache = new Map();
        this.capacity = capacity;
        this.head = null;
        this.tail = null;
    }
    
    get(key) {
        if (this.cache.has(key)) {
            let data = this.cache.get(key);
            this.removeNode(data);
            this.appendNode(data);
            return data.value;
        } else {
            return -1;
        }
    }
    
    put(key, value) {
        if (this.cache.has(key)) {
            let data = this.cache.get(key);
            data.value = value;
            this.removeNode(data);
            this.appendNode(data);
        } else if (this.cache.size < this.capacity) {
            let data = new Node(key, value);
            this.cache.set(key, data);
            this.appendNode(data);
        } else {
            let data = this.head;
            let oldKey = data.key;
            data.key = key;
            data.value = value;
            this.cache.delete(oldKey);
            this.cache.set(key, data);
            this.removeNode(data);
            this.appendNode(data);
        }
    }
    
    removeNode(node) {
        if (!node) {
            return;
        }
        
        if (!!node.prev) {
            node.prev.next = node.next;
        }
        
        if (!!node.next) {
            node.next.prev = node.prev;
        }
        
        if (this.head === node) {
            this.head = node.next;
        }
        
        if (this.tail === node) {
            this.tail = node.prev;
        }
    }
    
    appendNode(node) {
        if (!node) {
            return;
        }
        
        node.prev = this.tail;
        node.next = null;
        
        if (!!this.tail) {
            this.tail.next = node;
        }
        this.tail = node;
        
        if (!this.head) {
            this.head = node;
        }
    }
}

class Node {
    constructor(key, value, prev = null, next = null) {
        this.key = key;
        this.value = value;
        this.prev = prev;
        this.next = next;
    }
}

var lruCache = new LRUCache(2);
lruCache.put(1, 1);
lruCache.put(2, 2);
lruCache.get(1);
lruCache.put(3, 3);
lruCache.get(2);
lruCache.put(4, 4);
lruCache.get(1);
lruCache.get(3);
lruCache.get(4);

File: Google/146 LRU Cache.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace LeetCode
{
    class Program
    {
        static void Main(string[] args)
        {
            var cache = new LRUCache(2);
            cache.Put(2, 1);
            cache.Put(3, 2);
            cache.Get(3);
            cache.Get(2);
            cache.Put(4, 3);
            cache.Get(2);
            cache.Get(3);
            cache.Get(4);
        }
    }

    public class LRUCache
    {
        private Dictionary<int, Data> Cache;
        private int Capacity;
        private Data Head;
        private Data Tail;

        public LRUCache(int capacity)
        {
            this.Cache = new Dictionary<int, Data>(capacity);
            this.Capacity = capacity;
        }

        public int Get(int key)
        {
            if (this.Cache.ContainsKey(key))
            {
                var data = this.Cache[key];
                this.RemoveNode(data);
                this.AddNode(data);

                return data.Value;
            }
            else
            {
                return -1;
            }
        }

        public void Put(int key, int value)
        {
            if (this.Cache.ContainsKey(key))
            {
                var data = this.Cache[key];
                data.Value = value;
                this.RemoveNode(data);
                this.AddNode(data);
            }
            else if (this.Cache.Count < this.Capacity)
            {
                var data = new Data
                {
                    Key = key,
                    Value = value,
                    Prev = this.Tail,
                };

                this.Cache[key] = data;
                this.AddNode(data);
            }
            else
            {
                var data = this.Head;
                var oldKey = data.Key;
                data.Key = key;
                data.Value = value;
                this.Cache.Remove(oldKey);
                this.Cache.Add(key, data);

                this.RemoveNode(data);
                this.AddNode(data);
            }
        }

        private void RemoveNode(Data data)
        {
            if (data == null)
            {
                return;
            }

            if (data.Prev != null)
            {
                data.Prev.Next = data.Next;
            }

            if (data.Next != null)
            {
                data.Next.Prev = data.Prev;
            }

            if (this.Head == data)
            {
                this.Head = data.Next;
            }

            if (this.Tail == data)
            {
                this.Tail = data.Prev;
            }
        }

        private void AddNode(Data data)
        {
            if (data == null)
            {
                return;
            }

            data.Prev = this.Tail;
            data.Next = null;

            if (this.Tail != null)
            {
                this.Tail.Next = data;
            }
            this.Tail = data;

            if (this.Head == null)
            {
                this.Head = data;
            }
        }
    }

    class Data
    {
        public int Key { get; set; }
        public int Value { get; set; }
        public Data Prev { get; set; }
        public Data Next { get; set; }
    }
}
